### ISSUE TITLE
Support for RFC 6961 / Multiple Certificate Status Request Extension

### ISSUE THREAD

**Issue by joernheissler** [openssl/openssl#4449]
Hello!

I'd like to see support for RFC 6961 in OpenSSL.
Did anything happen after <https://mta.openssl.org/pipermail/openssl-dev/2016-February/005309.html>?

I could give it a try myself, but I'm not (yet) familiar with OpenSSL code nor with the inner workings of TLS.
So, anyone else working on this feature?

**Comment by mattcaswell** [comment]
No one is working on it AFAIK.

**Comment by FdaSilvaYY** [comment]
@joernheissler : you will get some inspiration and tips in this PR #1008

**Comment by DDvO** [comment]
There are still requests for implementing OCSP multi-stapling in OpenSSL.
OTOH, https://tools.ietf.org/html/rfc8446#section-4.4.2.1 meanwhile states that 
the status_request_v2 extension [RFC6961] is deprecated for TLS 1.3.

Would it still make sense to implement it for TLS 1.2, 
since it will take quite a while until (sufficiently) all TLS clients and servers support TLS 1.3?

If so, any hints (in addition to the link to #1008) how to do that?
I would not be surprised if others have solved the problem already somewhere.

**Comment by joernheissler** [comment]
I added some pieces of code here: https://github.com/joernheissler/openssl/tree/i4449_status_request_v2
But that's far from anything useful.

One problem is how to get the status documents into the TLS server. This might need changes to how TLS servers parse their certificate/key files. I'm not even sure if openssl has support for this in TLS1.3.

**Comment by DDvO** [comment]
Thanks a lot Jörn for providing that code of nearly 3 years ago.
Very nice that it already contains documentation of the support function.
Do you know how much the changes overlap/conflict with the current OpenSSL master?

Assuming that the server side is already present/implemented elsewhere,
how far does this work already on the client side? Did you test this somehow?

**Comment by joernheissler** [comment]
It's three years ago, I don't remember it much.
I assume there will be major conflicts with current code.

Don't think I ever tested it. Java got some support for this extension, but I believe it's not really complete too.

**Comment by DDvO** [comment]
>  I'm not even sure if openssl has support for this in TLS1.3.

Me neither. 

The current documentation does not explicitly mention support for multiple "status_request" extensions, 
while this may be just because it has not been updated to the current state of TLSv1.3 support.

The `NEWS.md` entry on it says:
```
    * Fully compliant implementation of RFC8446 (TLSv1.3) on by default
```
which AFAICS this should mean that the - apparently mandatory - feature is supported.

OTOH, looking at the code in `extensions_{clnt,srvr}.c` I have the impression that also for TLSv1.3,
OpenSSL so far does not support cert status entries for intermediate certs (which corresponds to multi-stapling).

@richsalz, can you clarify?

**Comment by richsalz** [comment]
I do not believe multi-stapling has widespread adoption on the public Internet.  Perhaps @davidben can comment if Chrome supports it.

**Comment by richsalz** [comment]
I also do not think that anyone will complain about the "fully compliant" wording in the internal documentation.  The RFC is not a legal document, and as the IETF is fond of saying within its working groups, "we are not the protocol police."  It's fine, OpenSSL is fully compliant, don't worry.

**Comment by DDvO** [comment]
> I do not believe multi-stapling has widespread adoption on the public Internet.

Multi-stapling would certainly a good thing to have. One of the reasons why it's not that widespread yet is because OpenSSL does not support it. See also
* https://pfeifferszilard.hu/2020/09/09/why-do-certificate-revocation-checking-mechanisms-never-work.html
* https://bugs.openjdk.java.net/browse/JDK-8046321
* https://crypto.stackexchange.com/questions/42814/is-ocsp-multi-stapling-supported-by-openssl

**Comment by DDvO** [comment]
> I also do not think that anyone will complain about the "fully compliant" wording in the internal documentation. The RFC is not a legal document, and as the IETF is fond of saying within its working groups, "we are not the protocol police." It's fine, OpenSSL is fully compliant, don't worry.

We don't need to speak in 'legal' terms here - we are not lawyers :wink: 
So, let's talk techn(olog)ically :smile:

So far, the current OpenSSL TLS implementation does not support cert status extensions for intermediate certs (which was typically called "OCSP multi-stapling" in v1.2). 
OpenSSL has already been supporting simple OCSP stapling for v1.2. Does it support this cert status extension for the end-entity cert in v1.3 mode, too?

**Comment by richsalz** [comment]
> Multi-stapling would certainly a good thing to have. One of the reasons why it's not that widespread yet is because OpenSSL does not support it. See also

I read those links.  Multi-staple would not solve the LetsEncrypt issue because *they did not revoke an intermediate.*
The last one just asks if OpenSSL supports it.

Perhaps there are some corner cases where it is useful, but nobody has presented them.  And I maintain it's not on the public Internet.

**Comment by DDvO** [comment]
Just to correct a wrong view of mine also in this thread: Meanwhile I got a hint from a colleague:
RFC 8446 section 9.2 ("Mandatory-to-Implement Extensions", https://tools.ietf.org/html/rfc8446#section-9.2)
does not include the "status_request" extension (for OCSP responses), so their support is not mandatory.
In particular, OCSP staples for intermediate CA certs (of client or server certs) are not required to be implemented even for TLS 1.3.

**Comment by kaduk** [comment]
> OpenSSL is fully compliant [with RFC 8446]

Technically we're not actually *fully* compliant -- we decided to continue to allow/offer (e.g.) DSA even for non-TLS-1.3 connections in order to preserve our API stability, despite there being a "MUST NOT" in the list at the end of https://tools.ietf.org/html/rfc8446#section-4.2.3 .  There might be one or two other points of divergence, but that's the one I remember the most.

**Comment by mattcaswell** [comment]
> The current documentation does not explicitly mention support for multiple "status_request" extensions,

We only support it on the leaf certificate. See:

https://github.com/openssl/openssl/blob/3a1ee3c1993c588a22cb80f1d0eb6237f83a1560/ssl/statem/extensions_clnt.c#L1397-L1406

**Comment by DDvO** [comment]
So support for those features should either be added soon, and as far as this does not happen,
it must be made clear in the documentation **before the 3.0 release** which of them are not supported.

**Comment by DDvO** [comment]
So what must be done at least is to _clearly document all points not compliant with RFC 8446_.
It should also be listed with optional TLS 1.3 features are supported and which not.
@richsalz, can you do that?

**Comment by richsalz** [comment]
> So, again, what must be done at least is to _clearly document all points not compliant with RFC 8446_.
> It should also be listed with optional TLS 1.3 features are supported and which not.
> @richsalz, can you do that?

I do not think this is important for 3.0, so I will not be spending any time on this.

**Comment by DDvO** [comment]
> > So, again, what must be done at least is to _clearly document all points not compliant with RFC 8446_.
> > It should also be listed with optional TLS 1.3 features are supported and which not.
> > @richsalz, can you do that?
> 
> I do not think this is important for 3.0, so I will not be spending any time on this.

You cannot claim on the one hand: `OpenSSL is fully compliant [with RFC 8446]`
and on the other not find the time to list at least those aspects that are (internally known to be) non-compliant.

**Comment by richsalz** [comment]
> You cannot claim on the one hand: OpenSSL is fully compliant [with RFC 8446]
and on the other not find the time to list at least those aspects that are (internally known to be) non-compliant.

https://xkcd.com/386/

**Comment by DDvO** [comment]
> It's three years ago, I don't remember it much.
> I assume there will be major conflicts with current code.
> 
> Don't think I ever tested it. Java got some support for this extension, but I believe it's not really complete too.

@joernheissler, I've just copied your branch and rebased it.

Luckily (or: sadly?) there was little overlap with other work on that corner of the TLS implementation,
the only (easy to solve) issues were that a couple of files such as `ssl_locl.h` meanwhile have been renamed
and that the error calls are done without function identifiers these days.
After a little more fiddling with a few lines in the code here and there, it compiles successfully.

BTW, you seem to be among the very few to exhibit good knowledge of Latin, 
using in an intermediate comment the first of your two commits the correct plural form of status:
```
// XXX This is the function which assembles the handshake message including all statūs.
```

**Comment by richsalz** [comment]
Have you tested that the major browsers still work?  You should test both with valid cert, and a revoked end-entity cert.

**Comment by DDvO** [comment]
The code is not complete, so does not even have alpha state yet.
So it would make little sense to throw its output at a real client or server at this stage.

**Comment by DDvO** [comment]
@joernheissler, I've meanwhile found that the test failures were simply due to a missing negation in
```
if (WPACKET_sub_memcpy_u24(pkt, resp, resp_len < 0 : 0 : resp_len))
```

Meanwhile I understand better what you meant by
> But that's far from anything useful.

Still your code does provide useful basic support functions
and good hints at which points the overall functionality needs to be extended.

**Comment by DDvO** [comment]
This feature request is still labeled "Post 1.1.1".

It looks like no-one here is actively working on it, and even if so, it would likely not fit to include it still into 3.0, right?
So I propose to replace the milestone by "Post 3.0.0".

**Comment by DDvO** [comment]
Assuming that someone would find the time to implement and submit at PR rather soon after 3.0 is out,
would this feature fit well for inclusion in 3.1?
Likely along with supporting the TLS 1.3 way of supporting multiple OCSP responses in both directions?

**Comment by t8m** [comment]
It depends on how invasive such implementation would be. But yeah most probably this is Post 3.0.0 material.

**Comment by richsalz** [comment]
Please not 3.0

It might be worth asking the IETF TLS working group (email tls@ietf.org) if there is much support for 6961.

**Comment by DDvO** [comment]
When adding the extended OCSP stapling capabilities, it will likely not be much difference
if we do this for TLS 1.2 (according to RFC 6961) and/or for TLS 1.3.

And there are sill many applications that will be stuck with 1.2 for a quite a while,
so I suppose it would be worth and not much extra effort to support 
both OCSP multi-stapling (6961) and the new "status_request" extension (1.3).

**Comment by HBrock** [comment]
@richsalz I would like to give you a little background regarding our motivation to ask for OCSP multi-stapling.

Like other vendors in the EU rail automation industry, we are currently implementing Subset-137 (https://www.era.europa.eu/sites/default/files/filesystem/ertms/ccs_tsi_annex_a_-_mandatory_specifications/set_of_specifications_3_etcs_b3_r2_gsm-r_b1/index083_-_subset-137_v100.pdf). Subset-137 requires certificate validation using OCSP, see Table 2 and section 6.2.4.6. Currently there is a discussion in standardization to require multi-stapling especially on the link from the track-side to the onboard units.

**Comment by richsalz** [comment]
So it's currently in discussion, not a standard.  Perhaps you should tell them that it isn't implemented and push back.

**Comment by HBrock** [comment]
Multi-stapling is already optional in the standard. The discussion is, if they will make it mandatory.
We informed them about lack of support. But we also see the advantages to use multi-stapling on this weak wireless link between train and track-side.

**Comment by richsalz** [comment]
>  But we also see the advantages to use multi-stapling on this weak wireless link between train and track-side.

Can you talk about the advantages?  I think that the issues are about the revocation of intermediates (how often, other lifecycle) and not the transport.  In the WebPKI, they don't worry about the lifecycle of intermediates, it would be helpful to know how/why this is different.

**Comment by HBrock** [comment]
> In the WebPKI, they don't worry about the lifecycle of intermediates, it would be helpful to know how/why this is different.

This is not a WebPKI use case. 
To be honest, I am not involved in the standardization at UNISIG. I only know that status checking for Intermediate CAs is already required. Using multi-stapling they want to reduce the number of additional sessions for querying the OCSP responder.

Thanks for the input. We will discuss that further internally.

**Comment by hariomsaxena25** [comment]
Hi All,

Can this request be taken care for implementation. Is there any code using which i can test this feature.

**Comment by filipdimitrovcom** [comment]
i all,

Unfortunately for me, I've ended up here :)

So I've got joernheissler branch and rebased it (sort of) to master.
Then I've created my own branch, and extended it to fully working code, based on his API functions.
Changes should affect only TLS 1.2, and should be fully compatible with allready written code that uses openssal lib.

I've tested it with only few openssl build options (on and off "no-ocsp", debug and release).
So obviously this needs more testing.
Also my code changes need serious code review by someone more involved into openssl project before wide use.
All openssl tests are working, and I've tested RFC 6961 with my own applications.

Internally I'm using one structure to hold data (the new one) for multiple and single responses.
So no matter the API used, ocsp responses can be transferred by both extensions.
I.e. clients of openssl lib may use old API, but if new extension is requested, it will be used.
And vice-versa, if user use new API functions, but old extension is requested, only first (leaf) ocsp will be sent by old extension.

TLS client can use new extension to request data in old format, this is supported (and tested - maybe).
However, if SSL_set_tlsext_status_type(ssl, TLSEXT_STATUSTYPE_ocsp_multi) is called in TLS client context, 
request to server will be always to use new format : CertificateStatusType = ocsp_multi.

Some problems and todos here :

No way for client to request both TLSEXT_STATUSTYPE_ocsp_multi and TLSEXT_STATUSTYPE_ocsp OCSP response from server.
Also server will not respond to both requests if received in one ClientHello, server will respond only to last found.
This present a problem for clients that want to support both variants in one connection attempt, but server support only one.

Some functions internally call SSL_set_tlsext_status_type(s, TLSEXT_STATUSTYPE_ocsp) like SSL_set_ct_validation_callback().
If TLSEXT_STATUSTYPE_ocsp_multi is desired, call SSL_set_tlsext_status_type() after SSL_set_ct_validation_callback().
    
Perhaps easyest way to solve both problems above, is to alter SSL_set_tlsext_status_type() to add bitmask flags TLSEXT_STATUSTYPE_ocsp_multi and TLSEXT_STATUSTYPE_ocsp.
If SSL_set_tlsext_status_type(..., TLSEXT_STATUSTYPE_nothing) is called, clear both flags.
Internally ssl_ctx_st::ext::status_type & ssl_connection_st::ext::status_type both become bitmasks.
Also server should prioritize TLSEXT_STATUSTYPE_ocsp_multi over TLSEXT_STATUSTYPE_ocsp if both are received in ClientHello.
Would this change be accepted by openssl community or new API should be added?

TODO:
======================
x. write tests for multiple status request extension in openssl test framework (sorry, I'm not a perl guy :)
x. test with DTLS (is this needed?)
x. test with unknwon/invalid extension type multiple status request before valid one and after valid one
x. test for memory leaks
x. update s_client & s_server with test/example for multiple status request extension
x. update documentation for new public functions and describe interoperability between old/single and new/multi ocsp requests.

Regards,
Filip

**Comment by DDvO** [comment]
Very nice @filipdimitrovcom that you are continuing with the work by @joernheissler!

Would be easier to discuss with the relevant folks here if you place it here as draft pull request.
Then the changes become directly visible and can be commented/discussed in full context.

**Comment by DDvO** [comment]
I did not have time to look at the details, but I suspect that this strongly overlaps with #20945 - please check.

**Comment by filipdimitrovcom** [comment]
I suppose API and internal structure can be reused or should be merged for both implementations.
However in TLS 1.2 OCSP responses are send as separate handshake packet (deprecated in 1.3) while in 1.3 OCSP response is attached to each certificate itself, as extension.
So send/receive mechanism should stay for 1.2.
Not sure how to process, currently I do not have the time to align it, and #20945 is not yet accepted and merged into main.

Regards,
Filip

**Comment by filipdimitrovcom** [comment]
Hello all,
I've converted public API and internal data holder for OCSP responses to be exactly like that of #20945.
This will make merging both PRs much easier and from user point of view, there will be no need to use different APIs depending on if TLS 1.2 or TLS 1.3 is used.
Also documentation is updated, a bit different from #20945, merge may not be needed, but someone should review which one should be used. 
As for tests, they are already written for #20945, so they may be reused.
I've also rebased to master and squash my commits to just one.
PR is also converted to non draft one.

Cheers,
Filip

**Comment by nhorman** [comment]
should this be closed as a duplicate of https://github.com/openssl/openssl/issues/12921

**Comment by t8m** [comment]
> should this be closed as a duplicate of #12921

No, however IMO this one (not #12921) would be closed by #20945

**Comment by DDvO** [comment]
Multi-stapling for server certs is multi-stapling for server certs is being contributed here: #20945

**Comment by DDvO** [comment]
No, the latter implements half of this FR.

**Comment by DDvO** [comment]
> should this be closed as a duplicate of #12921

I'd say that since #20945 covers server-side stapling, this FR asks for both side,
we can close this FR here when #20945 is done because the other (missing) half is covered by #12921.

**Comment by fwh-dc** [comment]
PR was merged, so I guess this can be closed? Please reopen if I’m wrong.
